### PR TITLE
remote.http: new component

### LINKED ISSUE #2270: Flow: discovery.http component
For feature parity, we should bring over the upstream Prometheus SD mechanism and introduce it as a Flow component.

### PR DESCRIPTION
**Pull request by rfratto**
Introduce a `remote.http` component which polls a specified URL. The response body of the request is exposed to other components as either a string or a secret, depending on the value of the `is_secret` argument.

This component acts as a generic form of Prometheus' HTTP SD, so it implements #2270.

While documenting this component, I noticed some of the shared blocks accidentally included component-specific detail. All of the shared blocks have been updated to make sure they're documenting the correct information at a generic level.

Closes #2270.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,11 @@ Main (unreleased)
 
 - riverfmt: Allow function arguments to persist across different lines. (@rfratto)
 
+- New Grafana Agent Flow components:
+
+  - `remote.http` polls an HTTP URL and exposes the response as a string or
```
Comment (mattdurham) [inline CHANGELOG.md:23]: Should this be exposes the `body` instead of the response? My mind goes to response codes.

Comment (rfratto) [reply CHANGELOG.md:23]: yep, good point

**Review by mattdurham — APPROVED** [review]
LGTM

File: docs/sources/flow/reference/components/remote.http.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+`poll_frequency` | `duration` | Frequency to poll the URL. | `"1m"` | no
+`poll_timeout` | `duration` | Timeout when polling the URL. | `"10s"` | no
+`is_secret` | `bool` | Whether the response body should be treated as a secret. | false | no
+
+When `remote.http` performs a poll operation, an HTTP `GET` request is made
+against the URL specified by the `url` argument. A poll is triggered by the
+following:
+
+* When the component first loads.
+* Every time the components arguments get re-evaluated.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/remote.http.md:42]:
```suggestion
* Every time the component's arguments get re-evaluated.
```

File: docs/sources/flow/reference/components/remote.http.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+### tls_config block
+
+The `tls_config` block configures TLS settings for connecting to HTTPS servers.
+
+{{< docs/shared lookup="flow/reference/components/tls-config-block.md" source="agent" >}}
+
+## Exported fields
+
+The following fields are exported and can be referenced by other components:
```
Comment (karengermond) [inline docs/sources/flow/reference/components/remote.http.md:109]:
```suggestion
The following field is exported and can be referenced by other components:
```

File: docs/sources/flow/reference/components/remote.http.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`content` | `string` or `secret` | The contents of the file. | | no
+
+If the `is_secret` argument was `true`, `content` is a secret type.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/remote.http.md:115]: Where is this `is-secret` argument of which you speak?  Oh, is this supposed to be just `secret`?

**Review by karengermond — APPROVED** [review]
LGTM!  A couple minor things.

File: CHANGELOG.md [hunk 5]
```diff
@@ -18,6 +18,11 @@ Main (unreleased)
 - Grafana Agent Flow: Add tracing instrumentation and a `tracing` block to
   forward traces to `otelcol` component. (@rfratto)
 
+- New Grafana Agent Flow components:
+
+  - `remote.http` polls an HTTP URL and exposes the response body as a string
+    or secret to other components. (@rfratto)
```
Comment (tpaschalis) [inline CHANGELOG.md:24]: Let's not forget to register the component on `all.go`

File: component/remote/http/http.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments control the remote.http component.
+type Arguments struct {
+	URL           string        `river:"url,attr"`
+	PollFrequency time.Duration `river:"poll_frequency,attr,optional"`
+	PollTimeout   time.Duration `river:"poll_timeout,attr,optional"`
```
Comment (tpaschalis) [inline component/remote/http/http.go:39]: These are marked as optional, but the default value makes the decoding fail. Was the intention to make them required?

File: component/remote/http/http.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+// UnmarshalRiver implements river.Unmarshaler.
+func (args *Arguments) UnmarshalRiver(f func(interface{}) error) error {
+	type arguments Arguments
+	if err := f((*arguments)(args)); err != nil {
+		return err
+	}
+
+	if args.PollFrequency == 0 {
+		return fmt.Errorf("poll_frequency must be greater than 0")
```
Comment (tpaschalis) [inline component/remote/http/http.go:62]: Should we check for negative durations too? They're valid River expressions, too. 
(This is true for some other components that use durations like GCFrequency on otelcol.exporter.prometheus)

File: component/remote/http/http.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+// Run starts the remote.http component.
+func (c *Component) Run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(c.nextPoll()):
+			c.poll()
+		case <-c.updated:
+			// no-op; force the next wait to be reread.
```
[the remark this answers is not in the document]
Comment (rfratto) [reply component/remote/http/http.go:132]: It forces the wait period to be reread so if the poll interval is set to 2h and you change it to 1m, you don't have to finish waiting the original 2h before the new interval takes effect

File: CHANGELOG.md (hunk 5 above)
Comment (rfratto) [reply CHANGELOG.md:24]: 🤦‍♂️ thanks, I can't believe I'm still forgetting to do that :)

File: component/remote/http/http.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tpaschalis) [reply component/remote/http/http.go:132]: Oohhh, I see now, thanks!

File: docs/sources/flow/reference/components/remote.http.md (hunk 4 above)
Comment (rfratto) [reply docs/sources/flow/reference/components/remote.http.md:115]: It's the last argument in the arguments block :)

File: component/remote/http/http.go (hunk 6 above)
Comment (rfratto) [reply component/remote/http/http.go:39]: Oops, I forgot to apply the defaults

**Pull request merged**